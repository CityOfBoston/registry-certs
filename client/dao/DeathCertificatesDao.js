// @flow

import DataLoader from 'dataloader';
import type { LoopbackGraphql } from '../lib/loopback-graphql';
import type { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import lookupDeathCertificateOrder from '../queries/lookup-death-certificate-order';
import type {
  DeathCertificateSearchResults,
  DeathCertificateOrder,
} from '../types';

export type DeathCertificateCache = { [id: string]: DeathCertificate };

export default class DeathCertificatesDao {
  loopbackGraphql: LoopbackGraphql;
  loader: DataLoader<string, ?DeathCertificate>;

  cacheFullQuery: string;
  cacheByPage: { [page: number]: DeathCertificateSearchResults };

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader(ids =>
      fetchDeathCertificates(loopbackGraphql, ids).then(a => [...a])
    );
  }

  get(id: string): Promise<?DeathCertificate> {
    return this.loader.load(id);
  }

  async search(
    fullQuery: string,
    page: number
  ): Promise<DeathCertificateSearchResults> {
    if (fullQuery === this.cacheFullQuery) {
      const results = this.cacheByPage[page];
      if (results) {
        return results;
      }
    } else {
      this.cacheFullQuery = fullQuery;
      this.cacheByPage = {};
    }

    const { query, startYear, endYear } = this.parseQuery(fullQuery);

    const results = await searchDeathCertificates(
      this.loopbackGraphql,
      query,
      page,
      startYear,
      endYear
    );

    this.cacheByPage[page] = results;

    results.results.forEach(cert => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }

  async lookupOrder(
    id: string,
    contactEmail: string
  ): Promise<?DeathCertificateOrder> {
    return lookupDeathCertificateOrder(this.loopbackGraphql, id, contactEmail);
  }

  parseQuery(
    fullQuery: string
  ): { query: string, startYear: ?string, endYear: ?string } {
    // match a 4-digit year, and optionally a second 4-digit year with a hypen or en-dash
    const yearRegexp = /(\d{4})\s*[-â€“]?\s*(\d{4})?/;

    let query;
    let startYear;
    let endYear;

    const match = fullQuery.match(yearRegexp);
    if (match) {
      query = fullQuery.replace(yearRegexp, '').trim();
      startYear = match[1];
      endYear = match[2] || match[1];
    } else {
      query = fullQuery;
      startYear = null;
      endYear = null;
    }

    return {
      query,
      startYear,
      endYear,
    };
  }
}
