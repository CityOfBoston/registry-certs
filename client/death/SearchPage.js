// @flow

import React from 'react';
import Head from 'next/head';
import Router from 'next/router';
import type { Context } from 'next';

import type { ClientDependencies } from '../page';

import type Cart from '../store/Cart';
import Nav from '../common/Nav';
import Pagination from '../common/Pagination';
import type { DeathCertificateSearchResults } from '../types';

import SearchResult from './search/SearchResult';

export type InitialProps = {|
  query: string,
  results: ?DeathCertificateSearchResults,
|};

export type Props = {
  ...InitialProps,
  cart: Cart,
};

type State = {
  query: string,
};

export default class IndexPage extends React.Component {
  props: Props;
  state: State;

  static async getInitialProps(
    ctx: Context<*>,
    { deathCertificatesDao }: ClientDependencies,
  ): Promise<InitialProps> {
    const { query } = ctx;

    let results = null;

    if (query.q) {
      results = await deathCertificatesDao.search(
        query.q,
        parseInt(query.page, 10) || 1,
      );
    }

    return {
      query: query.q || '',
      results,
    };
  }

  constructor(props: Props) {
    super(props);

    const { query } = props;

    this.state = {
      query,
    };
  }

  handleQueryChange = (ev: SyntheticInputEvent) => {
    this.setState({ query: ev.target.value });
  };

  handleSubmit = (ev: SyntheticInputEvent) => {
    const { query } = this.state;

    ev.preventDefault();
    Router.push(`/death?q=${encodeURIComponent(query)}`);
  };

  render() {
    const { results, cart } = this.props;
    const { query } = this.state;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificates</title>
        </Head>

        <Nav cart={cart} link="checkout" />

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title">Death Certificates</h1>
          </div>

          <form
            className="sf sf--md"
            acceptCharset="UTF-8"
            method="get"
            action="/death"
            onSubmit={this.handleSubmit}>
            <input name="utf8" type="hidden" value="✓" />

            <div className="sf-i">
              <input
                type="text"
                name="q"
                id="q"
                value={query}
                onChange={this.handleQueryChange}
                placeholder="Search…"
                className="sf-i-f"
                autoComplete="off"
              />
              <button className="sf-i-b" type="submit">Search</button>
            </div>
          </form>
        </div>

        {results && this.renderResults(results)}
        <div />

      </div>
    );
  }

  renderResults(results: DeathCertificateSearchResults) {
    // we want the query that was searched for
    const { query } = this.props;

    const start = 1 + (results.page - 1) * results.pageSize;
    const end = Math.min(start + results.pageSize - 1, results.resultCount);

    return (
      <div>
        <div className="p-a300 b--w">
          <div className="t--sans tt-u" style={{ fontSize: 12 }}>
            Showing {start}–{end} of {results.resultCount.toLocaleString()}{' '}
            results for “{query}”
          </div>
        </div>

        {results.results.map(certificate =>
          <SearchResult certificate={certificate} key={certificate.id} />,
        )}

        {results.resultCount > results.results.length &&
          this.renderPagination(results)}

        <div className="p-a300">
          Not finding what you’re looking for? Try refining your search or{' '}
          <a
            href="https://www.boston.gov/departments/registry/how-get-death-certificate"
            style={{ fontStyle: 'italic' }}>
            request a death certificate
          </a>.
        </div>
      </div>
    );
  }

  renderPagination({ page, pageCount }: DeathCertificateSearchResults) {
    const { query } = this.props;
    const makeHref = (p: number) => `/death?q=${query}&page=${p}`;

    return <Pagination page={page} pageCount={pageCount} hrefFunc={makeHref} />;
  }
}
