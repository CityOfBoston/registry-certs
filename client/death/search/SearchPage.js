// @flow

import React from 'react';
import Head from 'next/head';
import Router from 'next/router';

import {
  getDependencies,
  type ClientContext,
  type ClientDependencies,
} from '../../app';
import type { DeathCertificateSearchResults } from '../../types';

import AppLayout from '../../AppLayout';

import Pagination from '../../common/Pagination';

import SearchResult from './SearchResult';

type Props = {|
  query: string,
  page: number,
  results: ?DeathCertificateSearchResults,
|};

type State = {
  query: string,
};

export default class SearchPage extends React.Component<Props, State> {
  queryField: ?HTMLInputElement;

  static async getInitialProps(
    ctx: ClientContext,
    dependenciesForTest?: ClientDependencies
  ): Promise<Props> {
    const { query } = ctx;
    const { deathCertificatesDao } =
      dependenciesForTest || getDependencies(ctx);

    let q = query.q || '';
    let page = 1;

    let results = null;

    if (q) {
      page = parseInt(query.page, 10) || 1;

      results = await deathCertificatesDao.search(q, page);
    }

    return {
      query: q,
      page,
      results,
    };
  }

  constructor(props: Props) {
    super(props);

    const { query } = props;

    this.state = {
      query,
    };
  }

  submitSearch = (query: string) => {
    Router.push(`/death?q=${encodeURIComponent(query.trim())}`);
  };

  setQueryField = (queryField: ?HTMLInputElement) => {
    this.queryField = queryField;
  };

  handleQueryChange = (ev: SyntheticInputEvent<*>) => {
    const query: string = ev.target.value;
    this.setState({ query });
  };

  handleSubmit = (ev: SyntheticInputEvent<*>) => {
    ev.preventDefault();

    if (this.queryField) {
      this.queryField.blur();
    }

    const { query } = this.state;

    this.submitSearch(query);
  };

  render() {
    const { results, query: originalQuery } = this.props;
    const { query } = this.state;

    return (
      <AppLayout showNav>
        <div>
          <div className="b-c b-c--nbp">
            <Head>
              {originalQuery ? (
                <title>
                  Boston.gov — Death Certificates — Results for “{originalQuery}”
                </title>
              ) : (
                <title>Boston.gov — Death Certificates</title>
              )}
            </Head>

            <div className="sh sh--b0" style={{ paddingBottom: 0 }}>
              <h1 className="sh-title" style={{ marginBottom: 0 }}>
                {!results ? 'Death certificates' : 'Search results'}
              </h1>
            </div>

            {!results && this.renderIntro()}

            <div className="m-v300">
              <form
                className="sf sf--md"
                acceptCharset="UTF-8"
                method="get"
                action="/death"
                onSubmit={this.handleSubmit}
              >
                <input name="utf8" type="hidden" value="✓" />

                <div className="sf-i">
                  <input
                    aria-label="Search box"
                    aria-describedby="searchExamples"
                    type="text"
                    name="q"
                    ref={this.setQueryField}
                    value={query}
                    onChange={this.handleQueryChange}
                    placeholder="Search by full or partial name…"
                    className="sf-i-f"
                    autoComplete="off"
                  />
                  <button className="sf-i-b" type="submit">
                    Search
                  </button>
                </div>

                <div className="t--subinfo m-t200" id="searchExamples">
                  Examples: “j doe” “robert frost 1963” “johnson 1956-1957”
                </div>
              </form>
            </div>
          </div>

          {results && results.resultCount > 0 && this.renderResults(results)}
          {results && results.resultCount === 0 && this.renderNoResults()}
        </div>
      </AppLayout>
    );
  }

  renderIntro() {
    return (
      <div>
        <p className="t--intro">
          We have death certificates for anyone who died in Boston, or who
          listed Boston as their home.
        </p>

        <p className="t--info" style={{ fontStyle: 'normal' }}>
          To order a death certificate for someone, start by searching for their
          name:
        </p>
      </div>
    );
  }

  renderResults(results: DeathCertificateSearchResults) {
    // we want the query that was searched for
    const { query, page } = this.props;

    const start = 1 + (results.page - 1) * results.pageSize;
    const end = Math.min(start + results.pageSize - 1, results.resultCount);

    return (
      <div className="m-t700">
        <div className="b-c b-c--ntp b-c--nbp">
          <div className="t--sans tt-u p-v200" style={{ fontSize: 12 }}>
            Showing {start}–{end} of {results.resultCount.toLocaleString()}{' '}
            results for “{query}”
          </div>

          {results.results.map(certificate => (
            <SearchResult
              certificate={certificate}
              key={certificate.id}
              backUrl={`/death?q=${query}&page=${page}`}
            />
          ))}

          {results.resultCount > results.results.length &&
            this.renderPagination(results)}
        </div>

        <div className="b--g m-t700">
          <div className="b-c b-c--smv">
            <div className="t--info m-b300">
              Not finding who you’re looking for?
            </div>
            {this.renderHelp()}
          </div>
        </div>
      </div>
    );
  }

  renderNoResults() {
    return (
      <div className="b-c b-c--ntp">
        <div className="t--intro m-t700 m-b300">
          No results found for this search.
        </div>
        <div>{this.renderHelp()}</div>
      </div>
    );
  }

  renderHelp() {
    return (
      <ul className="ul t--subinfo" style={{ fontStyle: 'normal' }}>
        <li>
          We only have death records for people who either died in the City of
          Boston or who had Boston as their residence on their death
          certificate.
        </li>

        <li>
          This site has all deaths from 1956 on, but only a few records before
          then. You can{' '}
          <a href="https://www.boston.gov/departments/registry/how-get-death-certificate">
            request a death certificate from before 1956
          </a>.
        </li>

        <li>
          Search for both formal names and nicknames. For example, if there are
          no results for “Elizabeth,” try “Betty.”
        </li>

        <li>
          You can also search for a partial name. For example, “Jo” matches
          “Joe” and “Joseph.”
        </li>

        <li>
          If the decedent had a hyphenated first or last name, try searching for
          only one of the parts.
        </li>

        <li>
          Add a 4-digit year of death. You can also search for a range of death
          years, like “smith 1960–1965”
        </li>

        <style jsx>
          {`
            & > li {
              margin-top: 0.5em;
            }
          `}
        </style>
      </ul>
    );
  }

  renderPagination({ page, pageCount }: DeathCertificateSearchResults) {
    const { query } = this.props;
    const makeHref = (p: number) => `/death?q=${query}&page=${p}`;

    return (
      <div className="m-v300">
        <Pagination page={page} pageCount={pageCount} hrefFunc={makeHref} />
      </div>
    );
  }
}
