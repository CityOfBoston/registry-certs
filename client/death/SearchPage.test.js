// @flow
import Router from 'next/router';

import type { DeathCertificate, DeathCertificateSearchResults } from '../types';
import DeathCertificatesDao from '../dao/DeathCertificatesDao';

import { wrapSearchPageController } from './SearchPage';

import {
  TYPICAL_CERTIFICATE,
  PENDING_CERTIFICATE,
  NO_DATE_CERTIFICATE,
} from '../../fixtures/client/death-certificates';

jest.mock('next/router');
jest.mock('../dao/DeathCertificatesDao');

const TEST_DEATH_CERTIFICATES: DeathCertificate[] = [
  TYPICAL_CERTIFICATE,
  PENDING_CERTIFICATE,
  NO_DATE_CERTIFICATE,
];

const TEST_SEARCH_RESULTS: DeathCertificateSearchResults = {
  results: TEST_DEATH_CERTIFICATES,
  resultCount: 50,
  page: 0,
  pageSize: 5,
  pageCount: 10,
};

describe('getInitialProps', () => {
  let SearchPageController;
  let deathCertificatesDao;

  beforeEach(() => {
    deathCertificatesDao = new DeathCertificatesDao((null: any));

    const dependencies: any = { deathCertificatesDao };
    SearchPageController = wrapSearchPageController(
      () => dependencies,
      () => null
    );
  });

  it('works with no query', async () => {
    const initialProps = await SearchPageController.getInitialProps(
      ({ query: {} }: any)
    );

    expect(initialProps).toMatchSnapshot();
  });

  it('searches when given a query', async () => {
    deathCertificatesDao.search.mockReturnValue(TEST_SEARCH_RESULTS);

    const initialProps = await SearchPageController.getInitialProps(
      ({ query: { q: 'Monkey Joe' } }: any)
    );

    expect(initialProps).toMatchSnapshot();
    expect(deathCertificatesDao.search).toHaveBeenCalledWith('Monkey Joe', 1);
  });
});

describe('contentProps', () => {
  let contentProps;

  beforeEach(() => {
    const SearchPageController = wrapSearchPageController(
      () => ({}: any),
      p => {
        contentProps = p;
      }
    );

    new SearchPageController(({}: any)).render();
  });

  describe('submitSearch', () => {
    it('redirects to search for a query', () => {
      contentProps.submitSearch('Monkey Joe');
      expect(Router.push).toHaveBeenCalledWith('/death?q=Monkey%20Joe');
    });
  });
});
