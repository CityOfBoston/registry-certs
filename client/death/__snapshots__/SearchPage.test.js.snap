// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller getInitialProps searches when given a query 1`] = `
Object {
  "query": "Monkey Joe",
  "results": Object {
    "page": 0,
    "pageCount": 10,
    "pageSize": 5,
    "resultCount": 50,
    "results": Array [
      Object {
        "age": "5/1/1962",
        "deathDate": "3/4/2016",
        "deathYear": "2016",
        "firstName": "Bruce",
        "id": "000002",
        "lastName": "Banner",
        "pending": false,
      },
      Object {
        "age": "44 yrs. 2 mos. 10 dys",
        "deathDate": "4/15/2014",
        "deathYear": "2014",
        "firstName": "Logan",
        "id": "000001",
        "lastName": "Howlett",
        "pending": true,
      },
      Object {
        "age": "6",
        "deathDate": null,
        "deathYear": "2004",
        "firstName": "Monkey",
        "id": "000003",
        "lastName": "Joe",
        "pending": false,
      },
    ],
  },
}
`;

exports[`controller getInitialProps works with no query 1`] = `
Object {
  "query": "",
  "results": null,
}
`;
